driver-mapping/src/main/java/com/datastax/driver/mapping/Mapper.java

```diff
@@ -1101,49 +1194,110 @@ boolean isIncludedInQuery() {
             }
 
             @Override
-            void appendTo(Insert.Options usings) {
-                throw new UnsupportedOperationException("shouldn't be called");
+            void validate(QueryType qt, MappingManager manager) {
+                checkArgument(qt == QueryType.SAVE, "SaveNullFields option is only allowed in save queries");
             }
 
             @Override
-            void appendTo(Delete.Options usings) {
-                throw new UnsupportedOperationException("shouldn't be called");
+            boolean modifiesQueryString() {
+                return false;
             }
 
             @Override
-            void addToPreparedStatement(BoundStatement bs, int i) {
+            void modifyQueryString(BuiltStatement query) {
                 // nothing to do
             }
 
             @Override
-            void checkValidFor(QueryType qt, MappingManager manager) {
-                checkArgument(qt == QueryType.SAVE, "SaveNullFields option is only allowed in save queries");
+            int apply(BoundStatement bs, int currentIndex) {
+                return currentIndex;
             }
 
             @Override
-            boolean isIncludedInQuery() {
-                return false;
+            Object asCacheKey() {
+                // does not modify the query string
+                return Type.SAVE_NULL_FIELDS;
+            }
+
+            @Override
+            public boolean equals(Object other) {
+                return other == this ||
+                        (other instanceof SaveNullFields && this.saveNullFields == ((SaveNullFields) other).saveNullFields);
+            }
+
+            @Override
+            public int hashCode() {
+                return saveNullFields ? 1231 : 1237;
             }
         }
 
+        static class IfNotExists extends Option {
+            boolean ifNotExists;
+
+            IfNotExists(boolean ifNotExists) {
+                super(Type.IF_NOT_EXISTS);
+                this.ifNotExists = ifNotExists;
+            }
+
+            @Override
+            void validate(QueryType qt, MappingManager manager) {
+                checkArgument(qt == QueryType.SAVE, "IfNotExists option is only allowed in save queries");
```

Comment:
Could technically protocol version check here like we do for TTLs, although LWTs aren't tied to protocol version but CQL/C* version instead.